Improve selected menu item style

zecakeh opened the pull request:
I actually could do it directly with `Menu.Item`, I hadn't realized I could provide my own icon component.

![image](https://user-images.githubusercontent.com/76261501/105992774-ccfdfd80-60a5-11eb-8235-12c7cf4edb2d.png)

Tested on web Firefox Linux
Tested on native Android

tasn commented:
I didn't quite get your comment about being able to do it with `Menu.Item`. Could you elaborate?

Also, do we really have to manually set the styles? Doesn't paper handle anything themselves?

zecakeh commented:
> I didn't quite get your comment about being able to do it with Menu.Item. Could you elaborate?

It's just that in the Matrix Room I said I couldn't style the icon with Menu.Item, I was wrong.

> Also, do we really have to manually set the styles? Doesn't paper handle anything themselves?

Paper doesn't handle the "active" state for menu items so we have to handle that by ourselves. And we have to handle the icon colors manually because the default Paper behavior only looks nice with the default themes.

tasn commented:
Gotcha, so what you meant regarding `Menu.Item` is that you don't have to use the MUIIcon component, but intsead can just use the icon by name and style like this? If so, maybe worth doing? I just think that every `if` in the code is a bug waiting to happen. :)

zecakeh commented:
No, what I meant is that I thought I couldn't use `Menu.Item` at all if I wanted to style the icon. But I realized I can style the icon if I provide myself the MUIIcon component.

tasn commented:
Gotcha, so this patch already has your learnings applied?

tasn commented on src/widgets/MenuItem.tsx:
```diff
@@ -0,0 +1,24 @@
+      disabled={disabled}
+      icon={(icon) ? ({ color, size }) => <MaterialCommunityIcons name={icon} size={size} color={(active && !disabled) ? colors.activeIcon : color} /> : undefined}
+      theme={(active) ? { colors: { text: colors.active } } : undefined}
+      style={(active) ? { backgroundColor: colors.activeBackground } : undefined}
```
You should be merging your style with the style passed in props. Same I guess for the theme (though not as important). At the moment this widget fails where there's a style passed to it (the style passed will just override yours completely).

zecakeh replied:
The point was to use this style and theme, because if we want to have a consistent app we should style a component for every use case we have and reuse it as-is.

Maybe I should remove the theme and style props to make that clear.

tasn replied:
I understand this is the point, but it's currently in the type so people expect it to work.
Also, some styles may be useful still to set, so maybe just merge the styles. It's very easy to do, as the `style` prop accepts an array of styles that override each other...

zecakeh commented:
> Gotcha, so this patch already has your learnings applied?

Yes

zecakeh commented:
I rebased and modified as suggested. I had to change the style type available on props, otherwise we could have received an array of styles and it would have been more difficult to merge the styles.

zecakeh commented on src/widgets/MenuItem.tsx:
```diff
@@ -0,0 +1,26 @@
+import { useTheme } from "../theme";
+
+type PropsType = Omit<React.ComponentProps<typeof Menu.Item>, "icon" | "style"> & {
+  style?: ViewStyle;
```
You marked this as resolved, is this okay then?

tasn replied:
Yeah, because you already explained yourself before I made this comment.

zecakeh commented on src/widgets/MenuItem.tsx:
```diff
@@ -0,0 +1,26 @@
+    <Menu.Item
+      disabled={disabled}
+      icon={(icon) ? ({ color, size }) => <MaterialCommunityIcons name={icon} size={size} color={(active && !disabled) ? colors.activeIcon : color} /> : undefined}
+      theme={(active) ? { colors: { text: colors.active } } : undefined}
```
You said it was alright before, that's why I didn't touch it. I can merge it if you want

tasn replied:
I'm fine with not merging it, but at least Omit it from the passed props so people don't pass it.

zecakeh commented:
Done
